synapse/rest/client/v1/login.py

```diff
@@ -220,6 +266,11 @@ def _do_other_login(self, login_submission):
             login_submission,
         )
 
+        # If the identifier wasn't initially a m.id.user, ratelimiting already
+        # happened for this request.
+        if original_identifier_type == "m.id.user":
+            self._ratelimit_user(canonical_user_id)
```

Comment:
This feels a bit fragile tbh. Can we not move this up above the `validate_login`? Or always do it after?